### PR TITLE
[AST] Remove ParameterTypeFlags from ParenType and TupleType

### PR DESCRIPTION
**Pull request by hamishknight**
The last clients that relied on stashing parameter type flags on these types are now gone.

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

File: include/swift/AST/Types.h [hunk 1]
```diff
@@ -2123,6 +2115,11 @@ class ParameterTypeFlags {
   uint16_t toRaw() const { return value.toRaw(); }
 };
 
+/// A type that indicates how parameter flags should be handled in an operation
+/// that requires the conversion into a type that doesn't support them, such as
+/// tuples. They must either be dropped, or be enforced to not be present.
+enum class ParameterFlagHandling { Drop, AssertEmpty };
```
Comment (xedin) [inline include/swift/AST/Types.h:2121]: This is unsafe IHMO, be behavior should always be - assert if flags are present,  inside the solver `composeTuple` should only be called after appropriate checks have been performed. We really want to know about places where parameters with non-empty flags are converted into a tuple to figure out why the caller is not checking...

**Review by xedin — COMMENTED** [review]
Looks good to me, I only have one reservation and it's about `ParameterFlagHandling` and letting callers get away with dropping flags.

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

File: include/swift/AST/Types.h (hunk 1 above)
Comment (hamishknight) [reply include/swift/AST/Types.h]: Talked with Pavel offline, we're going to keep this as-is, but with `Drop` renamed to `IgnoreNonEmpty`. It turns out that most callers really do want to drop flags, and making this more convoluted runs the risk of callers just building the tuple themselves without explicitly calling out the fact they are dropping flags.

**Comment by hamishknight** [comment]
GitHub got the wrong timestamps for the force push, correct diff is https://github.com/apple/swift/compare/dab78c634d2adfc350f8f5085fce358040a4d2c8..69f59bbc6c8b2201b64b0100a8a62bffb79775e2

File: lib/APIDigester/ModuleAnalyzerNodes.cpp [hunk 2]
```diff
@@ -1601,8 +1601,9 @@ SwiftDeclCollector::constructTypeNode(Type T, TypeInitInfo Info) {
     // Still, return type first
     Root->addChild(constructTypeNode(Fun->getResult()));
 
-    auto Input = AnyFunctionType::composeTuple(Fun->getASTContext(),
-                                               Fun->getParams());
+    auto Input =
+        AnyFunctionType::composeTuple(Fun->getASTContext(), Fun->getParams(),
+                                      ParameterFlagHandling::IgnoreNonEmpty);
```
Comment (hamishknight) [inline lib/APIDigester/ModuleAnalyzerNodes.cpp:1606]: @nkcsgexi This preserves the current behavior of ignoring the parameter flags, but might we want to record parameter flags here for e.g `inout`, variadics, or ownership? AFAIK those would affect ABI. Should I file a bug?

Comment (nkcsgexi) [reply lib/APIDigester/ModuleAnalyzerNodes.cpp:1606]: Thanks for pinging, @hamishknight ! Wouldn't these flags be reflected in the mangled names of the function? We diagnose mangled name changes as ABI breakage.

Comment (hamishknight) [reply lib/APIDigester/ModuleAnalyzerNodes.cpp:1606]: Ah yes, in which case I guess this is fine

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@xedin Noticed that we didn't have logic to specifically reject inout arguments for tuple construction (previously we were relying on `TupleTypeElt::getType` producing `InOutType` to fail the match). Added logic to reject inout arguments up-front in 1bf954c61feaed20a7020f91739fc93e49a1c333, mind taking a look?

**Comment by xedin** [comment]
@hamishknight Looks great!

**Comment by hamishknight** [comment]
:ship:

**Pull request merged**